src/box/tx_memory.h

```diff
@@ -0,0 +1,212 @@
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+/*
+ * This file contains declaration class @a tx_stat that is common memory manager
+ * for every transaction manager and txns itself.
+ * Txns and transaction managers must allocate memory using methods @a tx_stat
+ * only because it make possible to monitor memory usage,
+ * and that is why every transaction manager must have its own memory manager
+ * derived from this one.
+ * It is also important to say that one should not create an instance of
+ * @a tx_stat, create instance of derivated memory manager instead.
+ */
+
+#include <trivia/util.h>
+#include "small/mempool.h"
+#include "histogram.h"
+struct txn;
+struct tuple;
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Allocations types for txns.
+ */
+enum TXN_ALLOC_TYPE {
+	TXN_ALLOC_STMT = 0,
+	TXN_ALLOC_SVP = 1,
+	TXN_ALLOC_USER_DATA = 2,
+	TXN_ALLOC_REDO_LOG = 3,
+	TXN_ALLOC_TRIGGER = 4,
+	TXN_ALLOC_TIMER = 5,
+	TXN_ALLOC_JOURNAL_ENTRY = 6,
+	TXN_ALLOC_MAX = 7
+};
+
+/**
+ * String appearance of enum TXN_ALLOC_TYPE
+ */
+extern const char *TXN_ALLOC_TYPE_STRS[];
+
+/**
+ * Storage for statistics of one allocation type for txn.
+ */
+struct txn_stat_storage {
+	struct histogram *hist;
+	uint64_t total;
+};
+
+/**
+ * Memory manager itself.
+ */
+struct tx_memory_manager {
+	/** Number of allocation types. */
+	uint32_t alloc_max;
+	/** Number of registered txn. */
+	uint64_t txn_num;
+	/**
+	 * Pointer to array of @txn_stat_storage objects.
+	 * It is better to allocate it as a field of derived memory manager.
+	 */
+	struct txn_stat_storage *stats_storage;
+};
+
+/**
+ * Memory usage of one txn. Allocated on its region.
+ */
+struct txn_mem_used {
+	/**
+	 * This field is used only in debug mode to make sure that txn has
+	 * deallocated all the mempool allocations before it was deleted.
+	 */
+	uint64_t mempool_total;
+	/** Total memory used for every type of allocations. */
+	uint64_t total[];
+};
+
+/**
+ * @brief A wrapper over region_alloc.
+ * @param txn Owner of a region.
+ * @param size Bytes to allocate.
+ * @param alloc_type See TXN_ALLOC_TYPE.
+ * @note The only way to truncate a region of @a txn is to clear @a txn.
+ */
+void *
+tx_memory_region_alloc(struct tx_memory_manager *stat, struct txn *txn,
```

Comment:
Imho it's not ok to pass to every alloc function tx_memory_manager and allocation type explicitly. Did you consider adding callback or reference to allocators so that they utilize memory manager on each allocation automatically in case it is set? Problem with allocation type can be solved by introducing helpers like savepoint_allocate(). Or mb manage to bind sizeof(struct) to alloc_type so that to pass only one argument..